### PR TITLE
[search] Added fuzzy search (up to 2 errors).

### PR DESCRIPTION
**Pull request by ygorshenin**
@mgsergio PTAL
@mpimenov CC

### CODE REVIEW

File: search/geocoder.cpp [hunk 1]
```diff
@@ -331,6 +334,18 @@ size_t OrderCountries(m2::RectD const & pivot, vector<shared_ptr<MwmInfo>> & inf
   auto const sep = stable_partition(infos.begin(), infos.end(), intersects);
   return distance(infos.begin(), sep);
 }
+
+size_t GetMaxErrorsForToken(UniString const & token)
+{
+  bool const digitsOnly = all_of(token.begin(), token.end(), [](UniChar c) { return isdigit(c); });
```
Comment (mgsergio) [inline search/geocoder.cpp]: `all_of(begin(token), end(token), isdigit);`

**Comment by ygorshenin** [comment]
https://jira.mail.ru/browse/MAPSME-3002

File: base/base_tests/levenshtein_dfa_test.cpp [hunk 2]
```diff
@@ -93,4 +93,21 @@ UNIT_TEST(LevenshteinDFA_Smoke)
...
+    TEST(Accepts(dfa, "москва"), ());
+    TEST(Accepts(dfa, "масква"), ());
+    TEST(Accepts(dfa, "моска"), ());
+    TEST(Rejects(dfa, "иосква"), ());
+  }
+  {
+    LevenshteinDFA dfa("москва", 0 /* prefixCharsToKeep */, 1 /* maxErrors */);
+    TEST(Accepts(dfa, "москва"), ());
+    TEST(Accepts(dfa, "иосква"), ());
+    TEST(Rejects(dfa, "моксва"), ());
```
Comment (ygorshenin) [inline base/base_tests/levenshtein_dfa_test.cpp]: Yes, I do. But not now - it's simple, but requires some work.

File: search/categories_cache.cpp [hunk 3]
```diff
@@ -52,29 +36,15 @@ CBV CategoriesCache::Load(MwmContext const & context)
...
-  vector<CBV> cbvs;
-
-  m_categories.ForEach([&](strings::UniString const & key, uint32_t const type) {
-    params.m_tokens[0][0] = key;
-    params.m_types[0][0] = type;
-
-    CBV cbv(RetrieveAddressFeatures(context, m_cancellable, params));
-    if (!cbv.IsEmpty())
-      cbvs.push_back(move(cbv));
+  m_categories.ForEach([&](uint32_t const type) {
```
Comment (ygorshenin) [inline search/categories_cache.cpp]: Done

File: search/feature_offset_match.hpp [hunk 4]
```diff
@@ -279,10 +287,11 @@ void MatchFeaturesInTrie(SearchTrieRequest<DFA> const & request,
   bool const categoriesMatched = MatchCategoriesInTrie(request, trieRoot, categoriesHolder);
 
   impl::OffsetIntersector<Filter, Value> intersector(filter);
-  ForEachLangPrefix(request, trieRoot,
-                    [&request, &intersector](TrieRootPrefix<Value> & langRoot, int8_t lang) {
-                      MatchInTrie(request.m_dfas, langRoot, intersector);
-                    });
+
+  ForEachLangPrefix(request, trieRoot, [&](TrieRootPrefix<Value> & langRoot, int8_t lang) {
```
Comment (ygorshenin) [inline search/feature_offset_match.hpp]: Done

File: search/geocoder.cpp [hunk 5]
```diff
@@ -219,15 +222,15 @@ MwmSet::MwmHandle FindWorld(Index const & index, vector<shared_ptr<MwmInfo>> con
   return handle;
 }
 
-strings::UniString AsciiToUniString(char const * s) { return strings::UniString(s, s + strlen(s)); }
+UniString AsciiToUniString(char const * s) { return UniString(s, s + strlen(s)); }
```
Comment (ygorshenin) [inline search/geocoder.cpp]: No need in the second const - I'm not planning to reassign pointer. You write
```c++
void Foo(int a, double b);
```
not
```c++
void Foo(int const a, double const b);
```
right? Here is the same thing.

**Comment by ygorshenin** [comment]
@mgsergio PTAL

**Comment by mgsergio** [comment]
LGTM

**Pull request merged**